lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -166,15 +165,16 @@ defmodule Mix.Compilers.Elixir do
       Agent.stop(pid, :normal, :infinity)
     else
       {:ok, _, _} ->
-        Agent.cast pid, fn {modules, sources, warnings} ->
+        Agent.get pid, fn {modules, sources, _} ->
           write_manifest(manifest, modules, sources, dest, timestamp)
-          {modules, sources, warnings}
+          {:ok, warning_diagnostics(sources)}
+        end
+      {:error, errors, _} ->
+        Agent.get pid, fn {_, sources, _} ->
+          error_diagnostics = Enum.map(errors, &(diagnostic(&1, :error)))
+          {:error, warning_diagnostics(sources) ++ error_diagnostics}
         end
```

Comment:
It is weird to send "errors" to the agent only to return it. We should probably rewrite this as:

```elixir
  errors = Enum.map(errors, &diagnostic(&1, :error))
  warnings = Agent.get(pid, fn {_, sources, _} -> warning_diagnostics(sources) end)
  {:error, warnings ++ errors}
```